internal/btf/core.go

```diff
@@ -197,35 +157,84 @@ func (k coreKind) String() string {
 	}
 }
 
-func coreRelocate(local, target *Spec, relos CORERelos) (COREFixups, error) {
+// ApplyRelocations calculates the difference in types between local
+// and target and adjusts insns to match.
+//
+// Passing a nil target will relocate against the running kernel. insns are
+// modified in place.
+func ApplyRelocations(local, target *Spec, insns asm.Instructions) error {
+	var relos []*CORERelocation
+	var reloInsns []*asm.Instruction
+	iter := insns.Iterate()
+	for iter.Next() {
```

Comment:
Could coreRelocate somehow be refactored to operate on a single Instruction? That way, getting relo from metadata, calculating the fixup and applying it can be done in a single loop without copying much. Not speaking about performance, but it might shorten this function and make it easier to follow. I haven't fully understood `coreRelocate()` though, just a suggestion.

Reply:
Unfortunately that doesn't work. coreRelocate splits all relocations into per-type relocations, and then uses the per-type relocations as "constraints" to find the best target type. That isn't possible when looking at a single instruction (and therefore relocation) in isolation.

Reply:
Okay, is this due to the way CO-RE is supposed to work, or just a decision that was made while doing the initial implementation?

Reply:
My initial implementation operated on a single instruction / relocation, and then failed some of the libbpf tests. As a result I changed the approach to what we have right now. It's not the same as what libbpf does (I think) but I'm fairly certain that only looking at a single relocation doesn't work.

Reply:
I'm curious to know why this is, but I might need to further my understanding of CO-RE first. We'll leave this as-is now.